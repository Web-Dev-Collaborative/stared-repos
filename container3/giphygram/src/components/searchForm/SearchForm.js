import React from 'react';
import PropTypes from 'prop-types';
import './searchForm.css';

const inputMaxLength = 512;

export class SearchForm extends React.Component {
  static propTypes = {
    query: PropTypes.string,
    onSearchSubmit: PropTypes.func,
    onSearchUpdate: PropTypes.func,
    onSearchReset: PropTypes.func,
  };

  static defaultProps = {
    query: '',
    onSearchSubmit: () => {},
    onSearchUpdate: () => {},
    onSearchReset: () => {},
  };

  onQueryChange = (event) => {
    const { onSearchUpdate } = this.props;
    const query = event.target.value;
    onSearchUpdate(query);
  };

  onSearchSubmit = (event) => {
    event.preventDefault();
    const { onSearchSubmit, query } = this.props;
    // Don't fire onSearchSubmit callback when search query is empty.
    if (query) {
      onSearchSubmit(query);
    }
  };

  onSearchReset = (event) => {
    event.preventDefault();
    const { onSearchReset } = this.props;
    onSearchReset();
  };

  render() {
    const { query } = this.props;

    const resetElement = query && query.length ? (
      <div className="input-group-append">
        <button
          className="btn btn-light search-reset"
          type="button"
          onClick={this.onSearchReset}
          title="Reset search"
        >
          <span className="oi oi-x" />
        </button>
      </div>
    ) : null;

    return (
      <form className="form" onSubmit={this.onSearchSubmit}>
        <div className="input-group">
          <input
            className="form-control search-input"
            type="search"
            placeholder="Search for GIFs"
            value={query}
            onChange={this.onQueryChange}
            spellCheck={false}
            autoCapitalize="off"
            autoComplete="off"
            maxLength={inputMaxLength}
          />

          {resetElement}

          <div className="input-group-append">
            <button
              className="btn btn-dark search-submit"
              type="submit"
              onClick={this.onSearchSubmit}
              title="Search"
            >
              <span className="oi oi-magnifying-glass" />
            </button>
          </div>
        </div>
      </form>
    );
  }
}
