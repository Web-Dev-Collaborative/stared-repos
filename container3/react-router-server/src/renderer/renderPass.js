import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
