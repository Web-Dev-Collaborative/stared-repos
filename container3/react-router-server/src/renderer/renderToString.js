import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
