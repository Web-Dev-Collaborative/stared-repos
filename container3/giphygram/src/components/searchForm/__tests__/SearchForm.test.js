import React from 'react';
import renderer from 'react-test-renderer';
import { SearchForm } from '../SearchForm';

describe('SearchForm', () => {
  it('should be rendered correctly with empty query', () => {
    const tree = renderer
      .create(<SearchForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be rendered correctly with non-empty query', () => {
    const tree = renderer
      .create(<SearchForm query="kittens" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fire onSearchSubmit form callback', () => {
    // Mock search form parameters.
    const searchQuery = 'kittens';
    const onSearchSubmit = jest.fn();

    // Create test component instance.
    const testComponentInstance = renderer.create((
      <SearchForm query={searchQuery} onSearchSubmit={onSearchSubmit} />
    )).root;

    // Try to find submit button inside the form.
    const submitButtonInstance = testComponentInstance.findByProps({
      type: 'submit',
    });
    expect(submitButtonInstance).toBeDefined();

    // Since we're not going to test the button component itself
    // we may just simulate its onClick event manually.
    const eventMock = { preventDefault: jest.fn() };
    submitButtonInstance.props.onClick(eventMock);

    expect(onSearchSubmit).toHaveBeenCalledTimes(1);
    expect(onSearchSubmit).toHaveBeenCalledWith(searchQuery);
  });

  it('should not fire onSearchSubmit form callback when request is empty', () => {
    // Mock search form parameters.
    const searchQuery = '';
    const onSearchSubmit = jest.fn();

    // Create test component instance.
    const testComponentInstance = renderer.create((
      <SearchForm query={searchQuery} onSearchSubmit={onSearchSubmit} />
    )).root;

    // Try to find submit button inside the form.
    const submitButtonInstance = testComponentInstance.findByProps({
      type: 'submit',
    });
    expect(submitButtonInstance).toBeDefined();

    // Since we're not going to test the button component itself
    // we may just simulate its onClick event manually.
    const eventMock = { preventDefault: jest.fn() };
    submitButtonInstance.props.onClick(eventMock);

    expect(onSearchSubmit).not.toHaveBeenCalled();
  });

  it('should fire onSearchReset form callback', () => {
    // Mock search form parameters.
    const searchQuery = 'kittens';
    const onSearchReset = jest.fn();

    // Create test component instance.
    const testComponentInstance = renderer.create((
      <SearchForm query={searchQuery} onSearchReset={onSearchReset} />
    )).root;

    // Try to find reset button inside the form.
    const resetButtonInstance = testComponentInstance.findByProps({
      type: 'button',
    });
    expect(resetButtonInstance).toBeDefined();

    const eventMock = { preventDefault: jest.fn() };
    resetButtonInstance.props.onClick(eventMock);

    expect(onSearchReset).toHaveBeenCalledTimes(1);
  });

  it('should fire onChange form callback', () => {
    // Mock search form parameters.
    const initialSearchQuery = 'kittens';
    const updatedSearchQuery = 'dogs';
    const onSearchUpdate = jest.fn();

    // Create test component instance.
    const testComponentInstance = renderer.create((
      <SearchForm query={initialSearchQuery} onSearchUpdate={onSearchUpdate} />
    )).root;

    // Try to find search input inside the form.
    const searchInputInstance = testComponentInstance.findByProps({
      type: 'search',
    });
    expect(searchInputInstance).toBeDefined();

    const eventMock = {
      preventDefault: jest.fn(),
      target: {
        value: updatedSearchQuery,
      },
    };
    searchInputInstance.props.onChange(eventMock);

    expect(onSearchUpdate).toHaveBeenCalledTimes(1);
    expect(onSearchUpdate).toHaveBeenLastCalledWith(updatedSearchQuery);
  });
});
