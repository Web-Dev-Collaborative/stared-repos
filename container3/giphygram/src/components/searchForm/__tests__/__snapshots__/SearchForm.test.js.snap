// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should be rendered correctly with empty query 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <input
      autoCapitalize="off"
      autoComplete="off"
      className="form-control search-input"
      maxLength={512}
      onChange={[Function]}
      placeholder="Search for GIFs"
      spellCheck={false}
      type="search"
      value=""
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-dark search-submit"
        onClick={[Function]}
        title="Search"
        type="submit"
      >
        <span
          className="oi oi-magnifying-glass"
        />
      </button>
    </div>
  </div>
</form>
`;

exports[`SearchForm should be rendered correctly with non-empty query 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <input
      autoCapitalize="off"
      autoComplete="off"
      className="form-control search-input"
      maxLength={512}
      onChange={[Function]}
      placeholder="Search for GIFs"
      spellCheck={false}
      type="search"
      value="kittens"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-light search-reset"
        onClick={[Function]}
        title="Reset search"
        type="button"
      >
        <span
          className="oi oi-x"
        />
      </button>
    </div>
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-dark search-submit"
        onClick={[Function]}
        title="Search"
        type="submit"
      >
        <span
          className="oi oi-magnifying-glass"
        />
      </button>
    </div>
  </div>
</form>
`;
